### PR TITLE
add NOT expression

### LINKED ISSUE #597: Support NOT expression
**Summary**

Datafusion implements Not expression. It is used in constant folding. 

https://github.com/apache/arrow-datafusion/blob/db4f098d38993b96ce1134c4bc7bf5c6579509cf/datafusion/src/logical_plan/expr.rs#L103

https://github.com/apache/arrow-datafusion/blob/db4f098d38993b96ce1134c4bc7bf5c6579509cf/datafusion/src/optimizer/constant_folding.rs#L149

### PR DESCRIPTION
**Pull request by jyizheng**
## Summary

Support negation of boolean expression

## Changelog

- New Feature

Fixes #597 

## Test Plan

Unit Tests

### CODE REVIEW

**Comment by jyizheng** [comment]
@BohuTANG   This  PR is ready. Could you please review it?

**Comment by BohuTANG** [comment]
Ok, ping @sundy-li for a review

**Comment by sundy-li** [comment]
Ok, Better after #550 , it will be finished in 2 days.

**Comment by BohuTANG** [comment]
> Ok, Better after #550 , it will be finished in at most 2 days.

Ok, thanks @sundy-li 
Cc @jyizheng

**Comment by jyizheng** [comment]
@sundy-li  Could you please review this PR?

**Comment by sundy-li** [comment]
> @sundy-li Could you please review this PR?

Ok, If your pr is not ready, you can create a draft PR. When it's ready, you can click `ready for review` button.  Refer: https://github.blog/2019-02-14-introducing-draft-pull-requests/

**Comment by jyizheng** [comment]
> > @sundy-li Could you please review this PR?
> 
> Ok, If your pr is not ready, you can create a draft PR. When it's ready, you can click `ready for review` button. Refer: https://github.blog/2019-02-14-introducing-draft-pull-requests/

Got it. It is ready now.

File: common/datavalues/src/data_array_logic.rs [hunk 1]
```diff
@@ -40,4 +37,29 @@ impl DataArrayLogic {
...
+    fn data_array_negation(val: &DataColumnarValue) -> Result<DataArrayRef> {
+        match val {
+            DataColumnarValue::Array(array) => {
+                let arr = downcast_array!(array, BooleanArray)?;
+                Ok(Arc::new(
+                    common_arrow::arrow::compute::not(arr).map_err(ErrorCodes::from)?
+                ))
+            }
+            _ => Result::Err(ErrorCodes::BadDataValueType(format!(
+                "DataValue Error: Cannot do data_array Not, array:{:?}",
```
Comment (sundy-li) [inline common/datavalues/src/data_array_logic.rs:50]: I think DataColumnarValue::Constant case should be considered.


1. Using `val.to_array()` , return the value as `DataColumnarValue::Array` 
2. Creating a array with just one element, and apply the `compute::not`, then return the value as `DataColumnarValue::Constant`

The second option can have better performance.

Comment (sundy-li) [reply common/datavalues/src/data_array_logic.rs:50]: We can have the stateless tests like `select number from numbers(10) where not ( 5 )`

**Review by sundy-li — CHANGES_REQUESTED** [review]

File: common/datavalues/src/data_array_logic.rs (hunk 1 above)
Comment (jyizheng) [reply common/datavalues/src/data_array_logic.rs:50]: Thanks for the suggestions. The function returns Result<DataArrayRef>, is it possible to return value as DataColumnarValue::Constant?

Comment (sundy-li) [reply common/datavalues/src/data_array_logic.rs:50]: It's ok.

Comment (jyizheng) [reply common/datavalues/src/data_array_logic.rs]: I have changed the code. Could you review it again?

**Pull request merged**